<a href="https://colab.research.google.com/github/michalis0/MGT-502-Data-Science-and-Machine-Learning/blob/main/assignments/Part 1/Assignment_part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DSML investigation:
### You are part of the Suisse Impossible Mission Force, or SIMF for short. You need to uncover a rogue agent that is trying to steal sensitive information.

### Your mission, should you choose to accept it, is to find that agent before any classified information gets stolen. Good luck!


## Assignment part one

Airport security confiscated a laptop of a spy, which was later given to you. The laptop contains some documents that might be useful in the investigation process. You retrieve those documents and want to analyse them. Our intelligence shows that the person that we are looking for has visited the USA between Sept 2019 to Oct 2020. He is currently working undercover.


### Getting to know our data

We have retrieved from the laptop the following file:
- A list of suspects
- The flight records of these potential suspects



In [21]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [22]:
user_records = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%201/data/userRecords.csv")
travel_records = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%201/data/flightRecords.csv")

#### Shape of the data

Let's first check how many rows and columns (features) are in the user records

In [23]:
# Your code here
print(user_records.shape)

(957, 8)



#### Check out the first few rows
Print the first few rows of the user records dataset and check them. Note that to protect innocent people, the name of the suspects have been censored, those will be revealed once number of potential suspects decreases.


In [24]:
# Your code here
print(user_records)

     UserID first_name  last_name  birthday country marital_status  \
0    317991      Lu***    Al*****  29.07.03      AR              1   
1    241892     Ma****  Ro*******  10.11.99      AR              1   
2    303376     Fr****     Ac****  25.07.84      AR              1   
3    761992     Fr****      Go***  20.03.70      AR              1   
4    373318      Jo***      Ro***  05.10.71      AR              1   
..      ...        ...        ...       ...     ...            ...   
952  218415   Bu******   Mo******  22.01.01      ZA              0   
953  173906    Th*****   Mo******  06.11.61      ZA              0   
954  178685    Mi*****  Bu*******  29.05.53      ZA              0   
955  200865  Si*******     Na****  16.12.62      ZA              0   
956  102637    Bo*****     Na****  24.02.50      ZA              0   

     number_of_kids               employment  
0               2.0             Web Designer  
1               0.0                   Banker  
2               0.

#### Column/feature names
Print the list of columns in the user records dataset.

In [25]:
# Your code here
user_records.columns

Index(['UserID', 'first_name', 'last_name', 'birthday', 'country',
       'marital_status', 'number_of_kids', 'employment'],
      dtype='object')

**Q1. How many suspects are there in the original user records dataset?**

*Hint: count the number of unique user IDs*

954

In [26]:
user_records['UserID'].nunique()

954

#### Duplicates
Check if there are any duplicate rows in the user records dataset.

__Remark__: If there are two rows in the dataset that have the same value for __all of the columns__, we consider this as a duplicate.

In [27]:
# Your code here
print(user_records.duplicated().sum())

3


Show all the duplicated rows in the user records dataset.

__Hint:__ use the method `duplicated(keep=False)` to get all occurrences of the duplicated rows.

In [28]:
# Your code here
print(user_records[user_records.duplicated(keep=False)])

     UserID first_name last_name  birthday country marital_status  \
18   100684    Ma*****    Me****  17.06.76      AR            NaN   
19   100684    Ma*****    Me****  17.06.76      AR            NaN   
47   910382      Ni***    El****  21.03.75      AZ            NaN   
165  910382      Ni***    El****  21.03.75      AZ            NaN   
276  610518    An*****  An******  15.02.53      DK              1   
290  610518    An*****  An******  15.02.53      DK              1   

     number_of_kids               employment  
18              0.0  Administrative employee  
19              0.0  Administrative employee  
47              0.0  Administrative employee  
165             0.0  Administrative employee  
276             NaN                  Manager  
290             NaN                  Manager  


**Q2 How many rows are dropped if we remove the duplicated rows in the user records dataset?**

3


Drop the duplicate rows in the user records dataset.

In [30]:
# Your code here
user_records = user_records.drop_duplicates()
user_records.shape

(954, 8)

### Spies often use false identities with fabricated data.

#### This suggests that if certain data points occur with statistically improbable frequency, they might be fabricated.

In this part we are going to verify if there are birthdates that have been reused an unusual amount of times.

Let's first ensure that the birthday column in the user records dataset has the correct format (datetime):


In [33]:
# Check the data type of each feature
# Your code here
user_records.dtypes

,0
UserID,int64
first_name,object
last_name,object
birthday,datetime64[ns]
country,object
marital_status,object
number_of_kids,float64
employment,object


Convert the column `birthday` to datetime if necessary. Make sure to pass the correct format when parsing dates. In this case, the correct date format is `"%d.%m.%y"`.

In [32]:
# Your code here
user_records['birthday'] = pd.to_datetime(user_records['birthday'], format="%d.%m.%y")

Find the list of birthdays that appear more than once in the dataset:

In [43]:
# Your code here
birthday_counts = user_records['birthday'].value_counts()
duplicate_birthdays = birthday_counts[birthday_counts > 1]
print(duplicate_birthdays)

birthday
1992-06-01    4
1970-04-06    2
2054-03-26    2
1983-03-15    2
2060-06-23    2
2068-09-22    2
2057-06-25    2
2057-12-24    2
1991-12-02    2
2060-03-26    2
1999-09-12    2
1999-08-09    2
2068-10-09    2
2062-07-07    2
2055-01-08    2
1970-03-18    2
1975-01-28    2
1983-08-26    2
2067-07-24    2
1978-11-21    2
2057-12-18    2
2004-11-30    2
Name: count, dtype: int64


In [44]:
print(duplicate_birthdays.dtype)

int64


Let's count how many people have their birthday on the same day among the duplicated birthdays.

In [41]:
# Your code here
duplicate_birthdays_count = duplicate_birthdays.sum()
print(duplicate_birthdays_count)

46


**Q3. What is the most common birthday in the dataset?**

01.06.1992

**Q4. How many users have their birthdays on that day (the most common birthday)?**

4

In [45]:
user_records.dtypes

,0
UserID,int64
first_name,object
last_name,object
birthday,datetime64[ns]
country,object
marital_status,object
number_of_kids,float64
employment,object


We can consider that there is no suspicious discrepancy through birthdays.


#### We have reasons to believe that the suspect works very thoroughly, so he would make sure that the fabricated data about his identity is complete in order for his operation to run smoothly.
#### Therefore, we can exclude the people with incomplete information from the list of potential suspects.

Count how many null values are there in each column of the user records dataset. You can call `isnull()` and `sum()` to get a count of how many null values are there in each column.

In [46]:
# Your code here
user_records.isnull().sum()

,0
UserID,0
first_name,1
last_name,1
birthday,1
country,0
marital_status,102
number_of_kids,152
employment,1


If you encounter any null values in the dataset, be sure to exclude those entries from the suspect list (meaning removing any rows that contain at least one null value in any column).

In [47]:
# Your code here
user_records = user_records.dropna()
user_records.isnull().sum()

,0
UserID,0
first_name,0
last_name,0
birthday,0
country,0
marital_status,0
number_of_kids,0
employment,0


### Linking the user records to the flight records

We will now proceed to identify the users who were in the US from Sept 1st, 2019 to Oct 31st, 2020.

We will merge the flight records with the user records in order to ensure that users who have been unsuspected in the previous step are not considered.

In [48]:
# Check the user records dataset
user_records.head()

,UserID,first_name,last_name,birthday,country,marital_status,number_of_kids,employment
0,317991,Lu***,Al*****,2003-07-29,AR,1,2.0,Web Designer
1,241892,Ma****,Ro*******,1999-11-10,AR,1,0.0,Banker
2,303376,Fr****,Ac****,1984-07-25,AR,1,0.0,Kitchen staff
3,761992,Fr****,Go***,1970-03-20,AR,1,0.0,Teaching Assistant
4,373318,Jo***,Ro***,1971-10-05,AR,1,0.0,Administrative employee


In [49]:
travel_records = travel_records.rename(columns={'userID': 'UserID'})
# Check the flight records dataset
travel_records.head()

,flightName,departure,destination,date,userID
0,KX65826,AR,IT,20.01.10,317991
1,CL12336,IT,MT,16.06.10,317991
2,YN106988,MT,SE,08.05.11,317991
3,CB58289,SE,EE,08.11.11,317991
4,SX625101,EE,AZ,17.11.11,317991


[Merging](https://pandas.pydata.org/docs/user_guide/merging.html) the two datasets:

*Hint: Use the above questions to identify on what feature to merge the datasets*

The result should contain the features flightName, Departure, Arrival,  Date, UserID, first_name, last_name, etc.

In [51]:
# Your code here
travel_records = travel_records.rename(columns={'userID': 'UserID'})
merged_df = pd.merge(user_records, travel_records, on='UserID')
display(merged_df.head())

,UserID,first_name,last_name,birthday,country,marital_status,number_of_kids,employment,flightName,departure,destination,date
0,317991,Lu***,Al*****,2003-07-29,AR,1,2.0,Web Designer,KX65826,AR,IT,20.01.10
1,317991,Lu***,Al*****,2003-07-29,AR,1,2.0,Web Designer,CL12336,IT,MT,16.06.10
2,317991,Lu***,Al*****,2003-07-29,AR,1,2.0,Web Designer,YN106988,MT,SE,08.05.11
3,317991,Lu***,Al*****,2003-07-29,AR,1,2.0,Web Designer,CB58289,SE,EE,08.11.11
4,317991,Lu***,Al*****,2003-07-29,AR,1,2.0,Web Designer,SX625101,EE,AZ,17.11.11


Remember, our suspect operates meticulously, ensuring that there are no null values in any rows.

In [59]:
# Your code here
merged_df = merged_df.dropna()
merged_df.isnull().sum()

,0
UserID,0
first_name,0
last_name,0
birthday,0
country,0
marital_status,0
number_of_kids,0
employment,0
flightName,0
departure,0


Let's now identify the users who have **travelled to** the United States (US) (i.e. their destination is "US") between the 1st of september 2019 and the 31st of october 2020 (both dates included).

*Note that:*
 - *The travel date is not in datetime format. Again, use the date format `"%d.%m.%y"` to do the conversion to datetime.*
 - *The ``date`` field in travel record dataset represents both the departure and arrival date (i.e., all flights occur within the same day).*


In [60]:
# Your code here
merged_df['date'] = pd.to_datetime(merged_df['date'], format="%d.%m.%y")
US_travelers = merged_df[(merged_df['destination'] == 'US') & (merged_df['date'] >= '2019-09-01') & (merged_df['date'] <= '2020-10-31')]
display(US_travelers.head())
US_travelers.shape

,UserID,first_name,last_name,birthday,country,marital_status,number_of_kids,employment,flightName,departure,destination,date
246,468560,Ga******,Fe*******,2060-04-18,AR,1,2.0,Lawyer,MG891910,AR,US,2020-07-16
463,255830,Ch*******,Be****,2059-12-26,AT,1,1.0,Professor,OO3731010,CA,US,2020-04-10
958,861915,Ay***,Al*****,1978-02-26,AZ,0,1.0,Doctor,GU84573,ES,US,2020-02-10
2289,220420,Ma*****,So****,1984-12-23,BR,0,3.0,Web Designer,TF469810,IT,US,2020-03-17
2702,297653,Br***,Wo**,2068-09-22,CA,1,0.0,Lawyer,MX97137,FR,US,2020-07-23


(34, 12)

In [61]:
merged_df.shape

(19037, 12)

**Q5. How many suspects traveled to the US between the 1st of September 2019 and the 31st of October 2020?**

34

**Q6. Which users are included in the remaining list of suspects who traveled to the US between the 1st of September 2019 and the 31st of October 2020?** Choose from the answers on Moodle.

In [63]:
merged_df['first_name'].unique()

<function ndarray.sum>

In [66]:
print(US_travelers[['first_name', 'last_name']])

       first_name  last_name
246      Ga******  Fe*******
463     Ch*******     Be****
958         Ay***    Al*****
2289      Ma*****     So****
2702        Br***       Wo**
4089      Ra*****  He*******
4897        Sø***     Ma****
5349         Li**     Re****
7179        Ju***     Wi****
7451       Br****    De*****
7588    El*******      Ra***
7665        Er***  He*******
7913       Lu****      Bo***
8452       An****       Pa**
8979        Fi***   Mc******
9623         Pu**     Ra****
9990   Al********    Bi*****
12165      Ма****    Tu*****
14639        Ka**     Ha****
14875    Kr******      Di***
15972        In**    Al*****
16340     На*****    Во*****
16632      Га****   Мо******
16771      Ni****   Sv******
17222      Ni****    Ni*****
17644        Em**     Öz****
17894     Hü*****      Çe***
18116     Je*****      Pe***
18141       Na***      Sm***
18283       Li***       Di**
18312        Li**      Pa***
18362        Jo**     Fl****
18384        Lu**    Sa*****
18707       Ma

## [Data visualisation ](https://pandas.pydata.org/docs/user_guide/visualization.html)

 Suisse Impossible Mission Force is happy with the result and would like to have some visual aid to make a profile of the potential suspects.

Let's create a pie chart to visualize the employment distribution among the suspects who traveled to US between September 1st, 2019, and October 31st, 2020.

In [ ]:
# Your code here


**Q7. Which profession is the least represented among the suspects who traveled to US between September 1st, 2019, and October 31st, 2020?**
 Choose from the answers on Moodle.


## Journey length

The Suisse Impossible Mission Force suspects that the rogue agent is an anomaly within his declared employment category, particularly in terms of the number of days he spent at destinations compared to his peers.

To identify such deviations, we can employ boxplots for each employment category, allowing us to visually identify outliers in the data.

First, let's construct two tables: one for the dates of arrival and another for the dates of departure. Utilize the existing dataset to construct these tables based on the 'date', 'destination' and 'departure' columns (again note that 'date' field represents both departure and arrival date). Only consider arrival and departure flights whose date is between 1 September 2019 and 31 October 2020 (inclusive).

In [ ]:
# For this step we will need the information not only about the fights to the US,
# but also about the flights taken from the US
# Your code here


We can keep on both of these tables the user ID, employment, and departure date / arrival date.

In [ ]:
# Your code here


We can now join the two tables on user ID.

*Note that each suspect has at most one arrival flight to the US and one departure flight from the US during the investigation period. Therefore, when merging the arrival and departure tables, you should expect only one matched row per user ID.*

In [ ]:
# Your code here


We can now create a new feature called journey_length representing the difference between departure date and arrival date.

*Hint: Ensure that journey_length is of type numeric as the box plot in the next question requires a numeric type. To convert from datetime to number of days, you can use `dt.days`*

In [ ]:
# Your code here


*Sidenote: if you get negative journey length, what would be your explanation to that?*

We can now create the boxplots where the x-axis represents each profession and the y-axis corresponds to the journey length.

Based on these visualizations, we can look for the profession containing the most extreme outlier, as our intelligence suggests the suspect acts as an anomaly compared to their peers in that category.

In [ ]:
# Your code here


**Q8. Based on the boxplots, which profession shows the most pronounced outlier in journey length relative to others within the same profession (i.e., the individual whose journey length stands out the most compared to others in the same profession)?** Choose among the options on Moodle.

### SIMF is satisfied with your work but believes the current data is inconclusive.

Your investigation will continue when more intelligence comes in...

**Don't forget to complete the [moodle quiz](https://moodle.epfl.ch/mod/quiz/view.php?id=1280137) and submit your [code](https://moodle.epfl.ch/mod/assign/view.php?id=1282648) on Moodle before 23:59 Thursday 26.02.2026**